MyNUnit/MyNUnit/MyNUnit.cs

```diff
@@ -0,0 +1,327 @@
+                    exceptionDetails = ex.GetBaseException().Message;
+
+                    testPassed = exceptionType == testAttribute?.ExpectedException;
+                }
+            }
+            catch (Exception testExecEx)
+            {
+                exceptionDetails = testExecEx.Message;
+                exceptionType = testExecEx.GetType();
+            }
+            finally
+            {
+                foreach (var teardownMethod in methodSet.TearDownFunctions)
+                {
+                    try
+                    {
+                        teardownMethod.Invoke(classInstance, null);
+                    }
+                    catch (Exception ex)
+                    {
+                        throw new Exception($"Error in Teardown method {teardownMethod.Name}: {ex}");
+                    }
+                }
+            }
+
+            return new TestResult(testMethod.Name, testPassed, exceptionType, exceptionDetails, null, timer.Elapsed);
+        });
+    }
+
+    /// <summary>
+    /// Retrieves a list of file paths to assemblies from the specified directory.
+    /// </summary>
+    /// <returns>List of file paths to the assemblies.</returns>
+    private IEnumerable<string> FetchAssemblyFilePaths(string assemblyPath)
+        => Directory.GetFiles(assemblyPath, "*.dll", SearchOption.AllDirectories)
+            .Where(file => !file.Contains("MyNUnit.dll"))
+            .Select(Path.GetFullPath);
+
+    /// <summary>
+    /// Loads assemblies from the specified paths.
+    /// </summary>
+    /// <returns>List of loaded assemblies.</returns>
+    private List<Assembly> LoadAssemblies(string assemblyPath)
+    {
+        var assemblyPaths = FetchAssemblyFilePaths(assemblyPath);
+        var loadedAssemblies = new List<Assembly>();
+
+        foreach (var path in assemblyPaths)
+        {
+            var context = new AssemblyLoadContext(path, isCollectible: true);
+            var assembly = context.LoadFromAssemblyPath(path);
+
+            loadedAssemblies.Add(assembly);
+            assemblyContexts.Add(context);
+        }
+
+        return loadedAssemblies;
+    }
+
+    /// <summary>
+    /// Unloads all loaded assembly contexts.
+    /// </summary>
+    private void UnloadAssemblyContexts()
+    {
+        foreach (var context in assemblyContexts)
+        {
+            context.Unload();
+
+            for (var i = 0; i < 3; i++)
+            {
+                GC.Collect();
+                GC.WaitForPendingFinalizers();
+            }
+        }
+
+        assemblyContexts.Clear();
+    }
+
+    /// <summary>
+    /// Retrieves all classes exported from the assemblies.
+    /// </summary>
+    /// <returns>List of class types.</returns>
+    private IEnumerable<Type> RetrieveClassesFromAssemblies(string assemblyPath)
+    {
+        return LoadAssemblies(assemblyPath)
+            .SelectMany(a => a.ExportedTypes)
+            .Where(t => t.IsClass);
+    }
```

Comment:
```suggestion
    private IEnumerable<Type> RetrieveClassesFromAssemblies(string assemblyPath) =>
        LoadAssemblies(assemblyPath)
            .SelectMany(a => a.ExportedTypes)
            .Where(t => t.IsClass);
```
можно написать короче